mlir/include/mlir/IR/OpDefinition.h

```diff
@@ -932,6 +932,10 @@ struct SingleBlock : public TraitBase<ConcreteType, SingleBlock> {
   }
   template <typename OpT = ConcreteType>
   enable_if_single_region<OpT> insert(Block::iterator insertPt, Operation *op) {
+    Block *body = getBody();
+    // Insert op before the block's terminator if it has one
+    if (insertPt == body->end() && body->hasTerminator())
+      insertPt = Block::iterator(body->getTerminator());
```

Comment:
This is not a safe: if there is an unregistered operation at the end of the current block it'll insert before.

I don't see a safe way of doing this actually.

Reply:
Before 0ac21e654f194a0f7c9f5afe42e11924c546f89e this was actually (somewhat?) safe. `SingleBlockImplicitTerminator<OpTy>` must be instantiated with an op type, which means that the terminator op must be known (registered?).

I'm wondering if it's better to just always insert at the end of the block. It would simplify the API: no special handling for terminators, it behaves like a `vector::push_back`.

Reply:
Just a nit: the terminator must be known when the verifier is passing, this may not be the case in the middle of a transformation or while building the operation.

Reply:
Changing the definition of `hasTerminator` so that it checks the last operation is definitely a terminator and keeping current semantics should work, right? (of course reverting the change to the assertion in this PR)

Reply:
It's not clear to me that it'll fix it: if you can't be sure if the last op is a terminator or not, I don't quite see how you can decide to insert before or after it?

Reply:
My point is: if it is definitely a terminator, do not break the code and insert before it; keep `vector::push_back` semantics otherwise

Reply:
This depends if the operation is registered in the MLIRContext, so we're creating a different behavior here, and avoiding this is the reason the "mighHaveTrait" exists.

Reply:
So do you guys think we should just be consistent with `vector::push_back` semantics and just blame the API user if the code is broken? I'm fine with that, just wanted to keep former semantics.

Reply:
The push_back approach has the advantage of being consistent in terms of behavior (there is no "gotcha" involved). If there is an API misuse here it'll blow up immediately and consistently.

Reply:
Makes total sense. I'll push a follow-up PR.